File: src/main/java/app/repository/_BaseRepository.java
package app.repository;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;

public class _BaseRepository {
  private static final String DB_URL = "jdbc:sqlite:";
  private static final String DB_NAME = "db.sqlite";

  public static <T> T connect(final Command<T> action) {
    final var db = DB_URL + DB_NAME;//new File(System.getProperty("user.dir"), DB_NAME).
    try (var connection = DriverManager.getConnection(db)) {
      return action.run(connection);
    } catch (SQLException ex) {
      ex.printStackTrace();
      return null;
    }
  }

  protected interface Command<T> {
    T run(Connection connection) throws SQLException;
  }
}

File: src/app/repository/_BaseRepository.kt
@file:JvmName("Database")

package app.repository

import java.sql.Connection
import java.sql.DriverManager
import java.sql.ResultSet
import java.sql.SQLException
import java.util.*


private const val DB_NAME = "db.sqlite"

@JvmField
var lastError: Throwable? = null

fun <T> connect(action: (Connection) -> T?): T? {
	val db = "jdbc:sqlite:$DB_NAME"
	try {
		val properties = Properties()
		properties.setProperty("foreign_keys", "true")
		DriverManager.getConnection(db, properties).use { connection ->
			lastError = null
			// connection.createStatement().execute("PRAGMA foreign_keys=ON;")
			return action(connection)
		}
	} catch (ex: SQLException) {
		lastError = ex
		ex.printStackTrace()
		return null
	}
}

fun scalarQuery(sql: String): String? =
	connect {
		val res: ResultSet = it.createStatement().executeQuery(sql)
		res.next()
		res.getString(1)
	}
